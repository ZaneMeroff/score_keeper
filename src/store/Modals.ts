// Decorators
import { Action, Module, Mutation, VuexModule } from 'vuex-class-modules'

@Module
export default class Modals extends VuexModule {

  addPlayersModal: boolean = false
  clearDataModal: boolean = false
  deletePlayersModal: boolean = false
  resetScoreModal: boolean = false
  scoreLimitsModal: boolean = false

  get getShowAddPlayersModal(): boolean {
    return this.addPlayersModal
  }

  get getShowClearDataModal(): boolean {
    return this.clearDataModal
  }

  get getShowDeletePlayersModal(): boolean {
    return this.deletePlayersModal
  }

  get getShowResetScoreModal(): boolean {
    return this.resetScoreModal
  }

  get getShowScoreLimitsModal(): boolean {
    return this.scoreLimitsModal
  }

  // ------------------------------------------
  //       add players modal visibility
  // ------------------------------------------
  
  @Mutation
  addPlayersModalVisibility(show: boolean) {
    this.addPlayersModal = show
  }
  
  @Action
  action_addPlayersModalVisibility(show: boolean) {
    this.addPlayersModalVisibility(show)
  }

  // ------------------------------------------
  //        clear data modal visibility
  // ------------------------------------------
  
  @Mutation
  clearDataModalVisibility(show: boolean) {
    this.clearDataModal = show
  }
  
  @Action
  action_clearDataModalVisibility(show: boolean) {
    this.clearDataModalVisibility(show)
  }

  // ------------------------------------------
  //  confirm delete players modal visibility
  // ------------------------------------------
  
  @Mutation
  deletePlayersModalVisibility(show: boolean) {
    this.deletePlayersModal = show
  }

  @Action
  action_deletePlayersModalVisibility(show: boolean) {
    this.deletePlayersModalVisibility(show)
  }

  // ------------------------------------------
  //    confirm reset score modal visibility
  // ------------------------------------------
  
  @Mutation
  resetScoreModalVisibility(show: boolean) {
    this.resetScoreModal = show
  }

  @Action
  action_resetScoreModalVisibility(show: boolean) {
    this.resetScoreModalVisibility(show)
  }

  // ------------------------------------------
  //        score limit modal visibility
  // ------------------------------------------
  
  @Mutation
  scoreLimitModalVisibility(show: boolean) {
    this.scoreLimitsModal = show
  }

  @Action
  action_scoreLimitModalVisibility(show: boolean) {
    this.scoreLimitModalVisibility(show)
  }
}