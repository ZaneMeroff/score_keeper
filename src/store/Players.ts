// Vue & Decorators
import Vue from 'vue'
import { Action, Module, Mutation, VuexModule } from 'vuex-class-modules'

// Types
import { Player, PlayerData } from '@/types/players'

@Module
export default class Players extends VuexModule {

    playerData: PlayerData = {}

    get getPlayerData() {
      return this.playerData
    }

    // ------------------------------------------
    //              create player
    // ------------------------------------------

    @Mutation
    createPlayer(payload: Player) {
      Vue.set(this.playerData, payload.id, payload)
    }

    @Action
    action_createPlayer(payload: Player) {
      this.createPlayer(payload)
    }

    // ------------------------------------------
    //         modify target player score
    // ------------------------------------------

    @Mutation
    setPlayerScore(payload: { id: string, score: number }) {
      this.playerData[payload.id].score = payload.score
    }

    @Action
    action_setPlayerScore(payload: { id: string, score: number }) {
      this.setPlayerScore(payload)
    }

    // ------------------------------------------
    //         modify target player name
    // ------------------------------------------

    @Mutation
    setPlayerName(payload: { id: string, name: string }) {
      this.playerData[payload.id].name = payload.name
    }

    @Action
    action_setPlayerName(payload: { id: string, name: string }) {
      this.setPlayerName(payload)
    }

    // ------------------------------------------
    //        zero scores for all players
    // ------------------------------------------

    @Mutation
    zeroScores() {
      for (const player in this.playerData) {
        this.playerData[player].score = 0
      }
    }

    @Action
    action_zeroScores() {
      this.zeroScores()
    }
}