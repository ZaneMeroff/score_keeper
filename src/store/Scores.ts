// Decorators
import { Action, Module, Mutation, VuexModule } from 'vuex-class-modules'

// Types
import { Score, ScoreData } from '@/types/scores'
import Vue from 'vue'

@Module
export default class Scores extends VuexModule {

    scoreData: ScoreData = {}

    get getScoreData() {
      return this.scoreData
    }

    @Mutation
    setScoreData(payload: { id: string, updatedScore: number }) {
      // create or update score data based on id
      Vue.set(this.scoreData, payload.id, payload.updatedScore)
    }

    
}