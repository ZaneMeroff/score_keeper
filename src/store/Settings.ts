// Decorators
import Vue from 'vue'
import { Action, Module, Mutation, VuexModule } from 'vuex-class-modules'

@Module
export default class Settings extends VuexModule {

  darkMode: boolean = true
  minScore: number = 0
  maxScore: number = 10
  showScoreLimitsModal: boolean = false
  // showMenuDrawer: boolean = false

  get getDarkMode() {
    return this.darkMode
  }

  get getMinScore() {
    return this.minScore
  }

  get getMaxScore() {
    return this.maxScore
  }

  get getShowScoreLimitsModal() {
    return this.showScoreLimitsModal
  }

  // get getShowMenuDrawer() {
  //   return this.showMenuDrawer
  // }
  
  // ------------------------------------------
  //             toggle dark mode
  // ------------------------------------------

  @Mutation
  setDarkMode() {
    this.darkMode = !this.darkMode
  }

  @Action
  action_setDarkMode() {
    this.setDarkMode()
  }

  // ------------------------------------------
  //            set score limits
  // ------------------------------------------

  @Mutation
  setScoreLimits(payload: { min: number, max: number }) {
    this.minScore = payload.min
    this.maxScore = payload.max
  }

  @Action
  action_setScoreLimits(payload: { min: number, max: number }) {
    this.setScoreLimits(payload)
  }

  // ------------------------------------------
  //       score limit modal visibility
  // ------------------------------------------
  
  @Mutation
  setScoreLimitModalVisibility(show: boolean) {
    this.showScoreLimitsModal = show
  }

  @Action
  action_setScoreLimitModalVisibility(show: boolean) {
    this.setScoreLimitModalVisibility(show)
  }

  // ------------------------------------------
  //         menu drawer visibility
  // ------------------------------------------
  
  // @Mutation
  // setMenuDrawerlVisibility(show: boolean) {
  //   this.showMenuDrawer = show
  // }

  // @Action
  // action_setMenuDrawerlVisibility(show: boolean) {
  //   this.setMenuDrawerlVisibility(show)
  // }
}