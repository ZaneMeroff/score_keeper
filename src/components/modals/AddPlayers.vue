<template>
  <div>

    <!-- add players modal -->
    <q-dialog v-model="showModal" persistent>
      <q-card class="q-card" :dark="stateSettings.getDarkMode">
        <q-card-section>

          <!-- header -->
          <div class="text q-mb-sm" :style="stateSettings.getDarkModeText">Add Players</div>
          <q-separator color="blue" class="q-mb-xl" inset></q-separator>
          
          <!-- number of players input -->
          <q-input  
            v-model.number="numOfPlayers" 
            style="font-size: 16px"
            type="number"
            :dark="stateSettings.getDarkMode" 
            :rules="[ val => !!val.toString() || '* Required' ]"
            @input="onInputChange"
          />
          
          <q-btn-group flat class="q-btn-group">
            <!-- add button -->
            <q-btn
              flat
              class="q-btn"
              color="green"
              size="md"
              :label="'add'"
              :disable="disabled"
              @click="handleSaveBtn"
            />
            <!-- cancel button -->
            <q-btn
              flat
              class="q-btn"
              color="red"
              size="md"
              :label="'cancel'"
              @click="handleCancelBtn"
            />
          </q-btn-group>
        </q-card-section>
      </q-card>
    </q-dialog>

    <!-- error dialog -->
    <q-dialog v-model="showError" persistent>
      <q-card class="q-card" :dark="stateSettings.getDarkMode">
        <q-card-section>
          <div class="q-mb-md">
            <!-- header -->
            <div class="text q-mb-sm" :style="stateSettings.getDarkModeText"><q-icon color="red" name="error"></q-icon> Player Rules</div>
            <q-separator color="blue" inset></q-separator>
            <!-- rule text -->
            <div class="text q-my-lg" :style="stateSettings.getDarkModeText">Total players cannot exceed 50</div>
          </div>
          <!-- ok button -->
          <q-btn
            flat
            color="green"
            size="md"
            style="width: 100%"
            :label="'ok'"
            @click="showError = false"
          />
        </q-card-section>
      </q-card>
    </q-dialog>

  </div>
</template>
<script lang="ts">

// Decorators
import { Component, Prop, Vue } from 'vue-property-decorator'

// Vuex
import { stateModals, statePlayers, stateSettings } from '@/store/index'

@Component
export default class AddPlayers extends Vue {
  
  @Prop(Boolean) private showModal!: boolean
  
  disabled: boolean = false
  numOfPlayers: number = 1
  showError: boolean = false
  
  stateModals = stateModals
  statePlayers = statePlayers
  stateSettings = stateSettings

  handleSaveBtn() {
    if (this.validateRules()) {
      this.statePlayers.action_createPlayers(this.numOfPlayers)
      this.setDefault()
      this.stateModals.action_addPlayersModalVisibility(false)
      this.$emit('close')
    } else {
      this.showError = true
    }
  }
  
  handleCancelBtn() {
    this.setDefault()
    this.stateModals.action_addPlayersModalVisibility(false)
    this.$emit('close')
  }

  validateRules() {
    if (Object.keys(this.statePlayers.getPlayerData).length + this.numOfPlayers > 50) return false
    else return true
  }
  
  onInputChange() {
    if (this.numOfPlayers >= 1) this.disabled = false
    else this.disabled = true
  }

  setDefault() {
    this.numOfPlayers = 1
  }
}
</script>
<style scoped>

  .q-card {
    width: 300px;
  }

  .text {
    color: #FFF;
    font-size: 16px;
    text-align: center;
  }

  .q-btn {
    width: 50%;
  }

  .q-btn-group {
    display: inline !important;
  }

  .q-item-section {
    align-items: center;
    display: flex;
    padding: 10px 0px 20px 0px;
  }

</style>