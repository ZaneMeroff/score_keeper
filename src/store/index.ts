import Vue from 'vue'
import Vuex from 'vuex'
import VuexPersistence from 'vuex-persist'
import localforage from 'localforage'

import Modals from './Modals'
import Players from './Players'
import Settings from './Settings'

Vue.use(Vuex)

const vuexLocal = new VuexPersistence({
  storage: localforage as any,
  asyncStorage: true,
  modules: [ 'Modals', 'Players', 'Settings' ],
});

export const store: any = new Vuex.Store({
  // strict: process.env.DEV === 'true',
  state: {},
  modules: {},
  plugins: [ vuexLocal.plugin ],
})

export const stateModals= new Modals({ store, name: 'Modals' })
export const statePlayers= new Players({ store, name: 'Players' })
export const stateSettings = new Settings({ store, name: 'Settings' })
