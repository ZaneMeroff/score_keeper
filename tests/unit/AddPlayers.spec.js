import { createLocalVue, shallowMount } from '@vue/test-utils'
import AddPlayers from '@/components/modals/AddPlayers.vue'
import Vuex from 'vuex'
import Quasar, * as All from 'quasar'

import { store } from '@/store/index'
import Vue, { ComponentOptions, VueConstructor } from 'vue';

// --------- Vuex setup ------------

const localVue = createLocalVue()
localVue.use(Vuex, Quasar, {components: All, directives: All, plugins: All})

// ---------------------------------

describe('AddPlayers', () => {

  describe('snapshots', () => {

    it('should render with default html', () => {
      const component = shallowMount(AddPlayers, { localVue })
      
      expect(component.element).toMatchSnapshot()
    })
  })

  describe('props', () => {

    describe('showModal', () => {

      it('should accept a boolean', () => {
        const propsData = { showModal: true }
        const component = shallowMount(AddPlayers, { localVue, propsData })
        
        expect(component.vm.$props.showModal).toEqual(propsData.showModal)
      })
    })
  })

  describe('data', () => {

    describe('disabled', () => {

      it('should have a default value of false', () => {
        const propsData = { showModal: true }
        const component = shallowMount(AddPlayers, { localVue, propsData })
        
        expect(component.vm.$data.disabled).toEqual(false)
      })

      it('should accept a boolean', () => {
        const propsData = { showModal: true }
        const component = shallowMount(AddPlayers, { localVue, propsData })
        const newValue = true

        component.setData({ disabled: newValue })

        expect(component.vm.$data.disabled).toEqual(newValue)
      })
    })

    describe('numOfPlayers', () => {

      it('should have a default value of 1', () => {
        const propsData = { showModal: true }
        const component = shallowMount(AddPlayers, { localVue, propsData })
        
        expect(component.vm.$data.numOfPlayers).toEqual(1)
      })

      it('should accept a number', () => {
        const propsData = { showModal: true }
        const component = shallowMount(AddPlayers, { localVue, propsData })
        const newValue = 5

        component.setData({ numOfPlayers: newValue })

        expect(component.vm.$data.numOfPlayers).toEqual(newValue)
      })
    })

    describe('showError', () => {

      it('should have a default value of false', () => {
        const propsData = { showModal: true }
        const component = shallowMount(AddPlayers, { localVue, propsData })
        
        expect(component.vm.$data.showError).toEqual(false)
      })

      it('should accept a boolean', () => {
        const propsData = { showModal: true }
        const component = shallowMount(AddPlayers, { localVue, propsData })
        const newValue = true

        component.setData({ showError: newValue })

        expect(component.vm.$data.showError).toEqual(newValue)
      })
    })
  })

  describe('methods', () => {
  
    describe('handleSaveBtn', () => {
      
      it('should call the correct functions if validateRules returns true', () => {
        // ************************************************
        // ************************************************
        // ***************  incomplete  *******************
        // ************************************************
        // ************************************************
        // ************************************************
      })

      it('should set showError to true if validateRules returns false', () => {
        // ************************************************
        // ************************************************
        // ***************  incomplete  *******************
        // ************************************************
        // ************************************************
        // ************************************************
      })
    })

    describe('handleCancelBtn', () => {
      
      it('should call setDefault', () => {
        const propsData = { showModal: true }
        const component = shallowMount(AddPlayers, { localVue, propsData })
        const spy = jest.spyOn(component.vm, 'setDefault')

        component.vm.handleCancelBtn()

        expect(spy).toHaveBeenCalledTimes(1)
      })

      it.only('should call action_addPlayersModalVisibility(false)', () => {
        const propsData = { showModal: true }
        const component = shallowMount(AddPlayers, { localVue, propsData })
        component.vm.handleCancelBtn()
        // ************************************************
        // ************************************************
        // ***************  incomplete  *******************
        // ************************************************
        // ************************************************
        // ************************************************
      })
    })

    describe('validateRules', () => {
      
      it('should return false if total players is > 50', () => {
        const propsData = { showModal: true }
        const component = shallowMount(AddPlayers, { localVue, propsData })

        component.setData({ numOfPlayers: 51 })
        expect(component.vm.$data.numOfPlayers).toEqual(51)
        expect(component.vm.validateRules()).toEqual(false)
      })

      it('should return true if total players is <= 50', () => {
        const propsData = { showModal: true }
        const component = shallowMount(AddPlayers, { localVue, propsData })

        component.setData({ numOfPlayers: 49 })
        expect(component.vm.$data.numOfPlayers).toEqual(49)
        expect(component.vm.validateRules()).toEqual(true)
      })
    })

    describe('onInputChange', () => {
      
      it('should set disabled to false if numOfPlayers >= 1', () => {
        const propsData = { showModal: true }
        const component = shallowMount(AddPlayers, { localVue, propsData })

        expect(component.vm.$data.numOfPlayers).toEqual(1)
        component.vm.setDefault()
        expect(component.vm.$data.disabled).toEqual(false)
      })

      it('should set disabled to true if numOfPlayers < 1', () => {
        const propsData = { showModal: true }
        const component = shallowMount(AddPlayers, { localVue, propsData })

        component.setData({ numOfPlayers: -1 })
        expect(component.vm.$data.numOfPlayers).toEqual(-1)
        component.vm.onInputChange()
        expect(component.vm.$data.disabled).toEqual(true)
      })
    })

    describe('setDefault', () => {
      
      it('should set numOfPlayers back to 1', () => {
        const propsData = { showModal: true }
        const component = shallowMount(AddPlayers, { localVue, propsData })
        
        component.setData({ numOfPlayers: 5 })
        expect(component.vm.$data.numOfPlayers).toEqual(5)
        component.vm.setDefault()
        expect(component.vm.$data.numOfPlayers).toEqual(1)
      })
    })
  })
})
