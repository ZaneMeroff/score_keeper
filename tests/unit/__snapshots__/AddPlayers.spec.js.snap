// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddPlayers snapshots should render with default html 1`] = `
<div>
  <q-dialog-stub
    persistent="true"
    position="standard"
  >
    <q-card-stub
      class="q-card"
      dark="true"
      tag="div"
    >
      <q-card-section-stub
        tag="div"
      >
        <div
          class="text q-mb-sm"
          style="color: rgb(255, 255, 255);"
        >
          Add Players
        </div>
         
        <q-separator-stub
          class="q-mb-xl"
          color="blue"
          inset="true"
        />
         
        <q-input-stub
          dark="true"
          rules="[Function]"
          style="font-size: 16px;"
          type="number"
          value="1"
        />
         
        <q-btn-group-stub
          class="q-btn-group"
          flat="true"
        >
          <q-btn-stub
            align="center"
            class="q-btn"
            color="green"
            flat="true"
            label="add"
            ripple="true"
            size="md"
          />
           
          <q-btn-stub
            align="center"
            class="q-btn"
            color="red"
            flat="true"
            label="cancel"
            ripple="true"
            size="md"
          />
        </q-btn-group-stub>
      </q-card-section-stub>
    </q-card-stub>
  </q-dialog-stub>
   
  <q-dialog-stub
    persistent="true"
    position="standard"
  >
    <q-card-stub
      class="q-card"
      dark="true"
      tag="div"
    >
      <q-card-section-stub
        tag="div"
      >
        <div
          class="q-mb-md"
        >
          <div
            class="text q-mb-sm"
            style="color: rgb(255, 255, 255);"
          >
            <q-icon-stub
              color="red"
              name="error"
              tag="i"
            />
             Player Rules
          </div>
           
          <q-separator-stub
            color="blue"
            inset="true"
          />
           
          <div
            class="text q-my-lg"
            style="color: rgb(255, 255, 255);"
          >
            Total players cannot exceed 50
          </div>
        </div>
         
        <q-btn-stub
          align="center"
          color="green"
          flat="true"
          label="ok"
          ripple="true"
          size="md"
          style="width: 100%;"
        />
      </q-card-section-stub>
    </q-card-stub>
  </q-dialog-stub>
</div>
`;
